extension/packages/hardhat/contracts/MetaMultiSigWallet.sol

```diff
@@ -0,0 +1,175 @@
+            isOwner[owner] = true;
+            emit Owner(owner, isOwner[owner]);
+        }
+        chainId = _chainId;
+    }
+
+    modifier onlySelf() {
+        require(msg.sender == address(this), "Not Self");
+        _;
+    }
+
+    function addSigner(address newSigner, uint256 newSignaturesRequired) public onlySelf {
+        require(newSigner != address(0), "addSigner: zero address");
+        require(!isOwner[newSigner], "addSigner: owner not unique");
+        require(newSignaturesRequired > 0, "addSigner: must be non-zero sigs required");
+        isOwner[newSigner] = true;
+        signaturesRequired = newSignaturesRequired;
+        emit Owner(newSigner, isOwner[newSigner]);
+    }
+
+    function removeSigner(address oldSigner, uint256 newSignaturesRequired) public onlySelf {
+        require(isOwner[oldSigner], "removeSigner: not owner");
+        require(newSignaturesRequired > 0, "removeSigner: must be non-zero sigs required");
+        isOwner[oldSigner] = false;
+        signaturesRequired = newSignaturesRequired;
+        emit Owner(oldSigner, isOwner[oldSigner]);
+    }
+
+    function updateSignaturesRequired(uint256 newSignaturesRequired) public onlySelf {
+        require(newSignaturesRequired > 0, "updateSignaturesRequired: must be non-zero sigs required");
+        signaturesRequired = newSignaturesRequired;
+    }
+
+    function getTransactionHash(
+        uint256 _nonce,
+        address to,
+        uint256 value,
+        bytes memory data
+    ) public view returns (bytes32) {
+        return keccak256(abi.encodePacked(address(this), chainId, _nonce, to, value, data));
+    }
+
+    function executeTransaction(
+        address payable to,
+        uint256 value,
+        bytes memory data,
+        bytes[] memory signatures
+    ) public returns (bytes memory) {
+        require(isOwner[msg.sender], "executeTransaction: only owners can execute");
+        bytes32 _hash = getTransactionHash(nonce, to, value, data);
+        nonce++;
+        uint256 validSignatures;
+        address duplicateGuard;
+        for (uint i = 0; i < signatures.length; i++) {
+            address recovered = recover(_hash, signatures[i]);
+            require(recovered > duplicateGuard, "executeTransaction: duplicate or unordered signatures");
+            duplicateGuard = recovered;
+            if (isOwner[recovered]) {
+                validSignatures++;
+            }
+        }
+
+        require(validSignatures >= signaturesRequired, "executeTransaction: not enough valid signatures");
+
+        (bool success, bytes memory result) = to.call{ value: value }(data);
+        require(success, "executeTransaction: tx failed");
+
+        emit ExecuteTransaction(msg.sender, to, value, data, nonce - 1, _hash, result);
+        return result;
+    }
+
+    function recover(bytes32 _hash, bytes memory _signature) public pure returns (address) {
+        return ECDSA.recover(MessageHashUtils.toEthSignedMessageHash(_hash), _signature);
```

Comment:
This line is changed because of `openZeppelin` v5. See previous version [here](https://github.com/scaffold-eth/se-2-challenges/blob/challenge-6-multisig/packages/hardhat/contracts/MetaMultiSigWallet.sol#L99)